lib/collection/src/collection/shard_transfer.rs

```diff
@@ -202,72 +198,56 @@ impl Collection {
-                shard_holder,
-                transfer.shard_id,
-                transfer.to,
-                activate_shard,
-                transfer.sync,
-            )
-            .await?;
-            log::debug!("proxy_promoted: {proxy_promoted}");
-        }
+        let is_resharding_transfer = transfer
+            .method
+            .map_or(false, |method| method.is_resharding());
 
-        // Should happen on receiving side
-        // Promote partial shard to active shard
-        // In case of resharding we manually promote the shard, and there are no partial shards
-        if self.this_peer_id == transfer.to
-            && !transfer
-                .method
-                .map_or(false, |method| method.is_resharding())
-        {
-            let shard_promoted =
-                transfer::driver::finalize_partial_shard(shard_holder, transfer.shard_id).await?;
-            log::debug!(
-                "shard_promoted: {shard_promoted}, shard_id: {}, peer_id: {}",
-                transfer.shard_id,
-                self.this_peer_id,
-            );
-        }
+        // Handle *source* replica
+        let src_replica_set = shard_holder.get_shard(&transfer.shard_id);
 
-        // Should happen on a third-party side
-        // Change direction of the remote shards or add a new remote shard
-        if self.this_peer_id != transfer.from {
-            // Normally we promote the shard to become active, in case of resharding we do not.
-            // For resharding we have multiple transfers in sequence, during which the shard should
-            // remain in the resharding state. Once all are done, the shard is manually promoted to active.
-            let state = if transfer
-                .method
-                .map_or(false, |method| method.is_resharding())
-            {
-                ReplicaState::Resharding
+        if let Some(replica_set) = src_replica_set {
+            if transfer.sync || is_resharding_transfer {
+                // If transfer is *sync* (or *resharding*), we *keep* source replica
+
+                if transfer.from == self.this_peer_id {
+                    // If current peer is *transfer-sender*, we need to unproxify local shard
+
+                    replica_set.un_proxify_local().await?;
+                }
             } else {
-                ReplicaState::Active
-            };
-            let remote_shard_rerouted = transfer::driver::change_remote_shard_route(
-                shard_holder,
-                transfer.shard_id,
-                transfer.to_shard_id.unwrap_or(transfer.shard_id),
-                transfer.from,
-                transfer.to,
-                state,
-                transfer.sync,
-            )
-            .await?;
-            log::debug!("remote_shard_rerouted: {remote_shard_rerouted}");
+                // If transfer is *not* sync (and *not* resharding), we *remove* source replica
+
+                if transfer.from == self.this_peer_id {
```

Comment:
Activation of target now depends on `replica_set.peer_state(&transfer.to).is_some()` condition. We should not remove source if target is not activated. 

Propose to handle target first and then return flag from there

Reply:
Good point. Will do.

Reply:
Done.